core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt

```diff
@@ -51,17 +63,158 @@ interface InfraExplorer {
+    val blockToPathProperties = mutableMapOf(block to pathProps)
+    val routes = blockInfra.routesOnBlock(rawInfra, block)
+    routes.forEach {
+        val incrementalPath = incrementalPathOf(rawInfra, blockInfra)
+        incrementalPath.extend(PathFragment(
+            mutableStaticIdxArrayListOf(it),
+            mutableStaticIdxArrayListOf(block),
+            containsStart = true,
+            containsEnd = false,
+            travelledPathBegin = location.offset.distance,
+            travelledPathEnd = Distance.ZERO
+        ))
+        val infraExplorer = InfraExplorerImpl(
+            rawInfra,
+            blockInfra,
+            mutableStaticIdxArrayListOf(block),
+            mutableStaticIdxArrayListOf(it),
+            incrementalPath,
+            blockToPathProperties
+        )
+        infraExplorers.add(infraExplorer)
+    }
+    return infraExplorers
+}
+
+private class InfraExplorerImpl(
+    private val rawInfra: RawInfra,
+    private val blockInfra: BlockInfra,
+    private var blocks: MutableStaticIdxArrayList<Block>,
+    private var routes: MutableStaticIdxArrayList<Route>,
+    private var incrementalPath: IncrementalPath,
+    private var blockToPathProperties: MutableMap<BlockId, PathProperties>,
+    private var currentIndex: Int = 0
+): InfraExplorer {
+
+    override fun getIncrementalPath(): IncrementalPath {
+        return incrementalPath
+    }
+
+    override fun getCurrentEdgePathProperties(offset: Offset<Block>, length: Distance?): PathProperties {
+        return PathPropertiesView(
+            blockToPathProperties[getCurrentBlock()]!!,
+            offset.cast(),
+            if (length == null)
+                Offset(blockInfra.getBlockLength(getCurrentBlock()).distance)
+            else
+                offset.plus(length).cast()
+        )
+    }
+
+    override fun getLastEdgeIdentifier(): EdgeIdentifier {
+        val currentAndRemainingBlocks = mutableStaticIdxArrayListOf<Block>()
+        for (i in currentIndex..<blocks.size) {
+            currentAndRemainingBlocks.add(blocks[i])
+        }
+        return EdgeIdentifierImpl(currentAndRemainingBlocks, routes.last())
+    }
+
+    override fun cloneAndExtendLookahead(): Collection<InfraExplorer> {
+        val infraExplorers = mutableListOf<InfraExplorer>()
+        val lastBlock = blocks.last()
+        val lastRoute = routes.last()
+        val lastBlockExit = blockInfra.getBlockExit(rawInfra, lastBlock)
+        val lastRouteExit = rawInfra.getRouteExit(lastRoute)
+        val nextRoutes =
+            if (lastBlockExit != lastRouteExit)
+                //Last route ends after last block: route should not change, but block should move forward
+                mutableStaticIdxArrayListOf(lastRoute)
+            else
+                rawInfra.getRoutesStartingAtDet(lastBlockExit)
+        nextRoutes.forEach {
+            val infraExplorer = this.clone()
+            (infraExplorer as InfraExplorerImpl).extend(it)
```

Comment:
(nitpicky, feel free to ignore)

```suggestion
            val infraExplorer = this.clone() as InfraExplorerImpl
            infraExplorer.extend(it)
```